core/trino-main/src/main/java/io/trino/metadata/HandleResolver.java

```diff
@@ -13,106 +13,66 @@
-                .add(EmptySplit.class)
-                .build());
+        classLoaders.put("system", HandleResolver.class.getClassLoader());
     }
 
-    public synchronized void addCatalogHandleClasses(String catalogName, Set<Class<?>> handleClasses)
+    public void registerClassLoader(PluginClassLoader classLoader)
     {
-        requireNonNull(catalogName, "catalogName is null");
-        requireNonNull(handleClasses, "handleClasses is null");
-        Set<Class<?>> existing = catalogHandleClasses.putIfAbsent(catalogName, ImmutableSet.copyOf(handleClasses));
-        checkState(existing == null, "Catalog is already registered: %s", catalogName);
-        for (Class<?> handleClass : handleClasses) {
-            allHandleClasses.add(handleClass);
-            handleClassIds.put(handleClass, classId(handleClass));
-        }
+        ClassLoader existingClassLoader = classLoaders.putIfAbsent(classLoader.getId(), classLoader);
+        checkState(existingClassLoader == null, "Class loader already registered: " + classLoader.getId());
     }
 
-    public synchronized void removeCatalogHandleClasses(String catalogName)
+    public void unregisterClassLoader(PluginClassLoader classLoader)
     {
-        Set<Class<?>> classes = catalogHandleClasses.remove(catalogName);
-        checkState(classes != null, "Catalog not registered: %s", catalogName);
-        for (Class<?> handleClass : classes) {
-            if (allHandleClasses.remove(handleClass, 1) == 1) {
-                handleClassIds.remove(handleClass);
-            }
-        }
+        boolean result = classLoaders.remove(classLoader.getId(), classLoader);
+        checkState(result, "Class loader not registered: " + classLoader.getId());
     }
 
-    public synchronized String getId(Object tableHandle)
+    @SuppressWarnings("MethodMayBeStatic")
+    public String getId(Object tableHandle)
     {
         Class<?> handleClass = tableHandle.getClass();
-        String id = handleClassIds.get(handleClass);
-        checkArgument(id != null, "Handle class not registered: " + handleClass.getName());
-        return id;
+        return classId(handleClass);
     }
 
-    public synchronized Class<?> getHandleClass(String id)
+    public Class<?> getHandleClass(String id)
     {
-        Class<?> handleClass = handleClassIds.inverse().get(id);
-        checkArgument(handleClass != null, "Handle ID not found: " + id);
-        return handleClass;
+        int splitPoint = id.lastIndexOf(':');
+        checkArgument(splitPoint > 1, "Invalid handle id: " + id);
+        String classLoaderId = id.substring(0, splitPoint);
+        String className = id.substring(splitPoint + 1);
+
+        ClassLoader classLoader = classLoaders.get(classLoaderId);
+        checkArgument(classLoader != null, "Unknown handle id: " + id);
+
+        try {
+            Class<?> handleClass = classLoader.loadClass(className);
+            if (handleClass != null) {
```

Comment:
Is this possible? It should throw instead of returning null

Reply:
It is not possible, but the old code was checking... I'll remove it